File: scripts/80_recompute_arima_rmse_on_var_window.R
# scripts/80_recompute_arima_rmse_on_var_window.R
suppressPackageStartupMessages({
  library(forecast)
})

rmse <- function(y_true, y_pred) {
  sqrt(mean((y_true - y_pred)^2))
}

expanding_one_step_arima <- function(y_train, y_test, order = c(0,1,0)) {
  history <- as.numeric(y_train)
  preds <- numeric(length(y_test))

  for (t in seq_along(y_test)) {
    fit <- forecast::Arima(history, order = order, include.drift = FALSE)
    preds[t] <- as.numeric(forecast::forecast(fit, h = 1)$mean[1])
    history <- c(history, as.numeric(y_test[t]))
  }
  preds
}

ROOT <- normalizePath(".")

eval_path <- file.path(ROOT, "output", "metadata", "eval_window.csv")
df_path   <- file.path(ROOT, "data", "processed", "var_dataset.csv")

if (!file.exists(eval_path)) stop("Missing output/metadata/eval_window.csv")
if (!file.exists(df_path))   stop("Missing data/processed/var_dataset.csv")

df <- read.csv(df_path)
df$date <- as.Date(df$date)

ew <- read.csv(eval_path)
ew$date <- as.Date(ew$date)

m <- merge(df, ew, by = "date")
m <- m[order(m$date), ]

y_train <- m$y[m$set == "train"]
y_test  <- m$y[m$set == "test"]

preds <- expanding_one_step_arima(y_train, y_test, order = c(0,1,0))
score <- rmse(y_test, preds)

out <- file.path(ROOT, "output", "tables", "rmse_baseline_arima_on_var_window.csv")
dir.create(dirname(out), recursive = TRUE, showWarnings = FALSE)
write.csv(data.frame(model="ARIMA(0,1,0) baseline", rmse=score, n_test=length(y_test)),
          out, row.names = FALSE)

cat("✅ wrote:", out, "\n")
cat(sprintf("ARIMA(0,1,0) RMSE on VAR test window: %.4f (n_test=%d)\n", score, length(y_test)))

File: scripts/80_recompute_arima_rmse_on_var_window.py
# scripts/80_recompute_arima_rmse_on_var_window.py
import sys
from pathlib import Path
import numpy as np
import pandas as pd
from statsmodels.tsa.arima.model import ARIMA

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

def rmse(y_true, y_pred) -> float:
    y_true = np.asarray(y_true, dtype=float)
    y_pred = np.asarray(y_pred, dtype=float)
    return float(np.sqrt(np.mean((y_true - y_pred) ** 2)))

def expanding_one_step_arima(y_train: pd.Series, y_test: pd.Series, order=(0,1,0)) -> np.ndarray:
    history = list(y_train.astype(float).values)
    preds = []
    for t in range(len(y_test)):
        model = ARIMA(history, order=order)
        fit = model.fit()
        yhat = float(fit.forecast()[0])
        preds.append(yhat)
        history.append(float(y_test.iloc[t]))
    return np.array(preds, dtype=float)

if __name__ == "__main__":
    eval_path = ROOT / "output" / "metadata" / "eval_window.csv"
    df_path   = ROOT / "data" / "processed" / "var_dataset.csv"  # 你后面 VAR 数据集建议固化成这个名字

    if not eval_path.exists():
        raise FileNotFoundError(f"Missing {eval_path}. Create it when you split VAR train/test.")
    if not df_path.exists():
        raise FileNotFoundError(f"Missing {df_path}. Create it when you build the VAR dataset (y + chosen x).")

    df = pd.read_csv(df_path, parse_dates=["date"])
    ew = pd.read_csv(eval_path, parse_dates=["date"])

    m = df.merge(ew, on="date", how="inner").sort_values("date")
    y_train = m.loc[m["set"] == "train", "y"]
    y_test  = m.loc[m["set"] == "test", "y"]

    preds = expanding_one_step_arima(y_train, y_test, order=(0,1,0))
    score = rmse(y_test.values, preds)

    out = ROOT / "output" / "tables" / "rmse_baseline_arima_on_var_window.csv"
    out.parent.mkdir(parents=True, exist_ok=True)
    pd.DataFrame({"model":["ARIMA(0,1,0) baseline"], "rmse":[score], "n_test":[len(y_test)]}).to_csv(out, index=False)

    print("✅ wrote:", out)
    print(f"ARIMA(0,1,0) RMSE on VAR test window: {score:.4f} (n_test={len(y_test)})")

